KAFKA-10000: Per-connector offsets topics (KIP-618)

C0urante opened the pull request:
Implements support for per-connector offsets topics as described in [KIP-618](https://cwiki.apache.org/confluence/display/KAFKA/KIP-618%3A+Exactly-Once+Support+for+Source+Connectors#KIP618:ExactlyOnceSupportforSourceConnectors-Per-connectoroffsetstopics).

Relies on changes from:
- https://github.com/apache/kafka/pull/11780

C0urante committed:
KAFKA-10000: Per-connector offsets topics

C0urante commented:
Given that all merge conflicts have been resolved and https://github.com/apache/kafka/pull/11780 has already been approved, marking this as ready for review.

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -1327,30 +1334,39 @@ public WorkerTask doBuild(Task task,
-                Map<String, Object> adminOverrides = adminConfigs(id.connector(), "connector-adminclient-" + id, config,
-                        sourceConfig, connectorClass, connectorClientConfigOverridePolicy, kafkaClusterId, ConnectorType.SOURCE);
-                topicAdmin = new TopicAdmin(adminOverrides);
+                topicAdminCreator.get();
```
I think this call to `get` could result in a leaked `TopicAdmin`, and thus `Admin`, because `offsetStoreForRegularSourceTask` doesn't guarantee to invoke `topicAdminCreator` (it calls it `topicAdminSupplier`) (when it does, the Admin will eventually we closed via `ConnectorOffsetBackingStore.stop`).

In any case I think the ownership of the `Admin` here (responsibility for ultimately closing it) is pretty unclear, but it's not clear to me why it _needs_ to be. So a comment to explain would be great, if a clean up of the logic is no possible.

C0urante replied:
This call won't result in a leak because the `AbstractWorkerSourceTask` class will always close its `TopicAdmin` during shutdown; see [AbstractWorkerSourceTask::close](https://github.com/C0urante/kafka/blob/fe675683b4671714d02c8e674bd36347d0c345e9/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTask.java#L306-L308). And, given the idempotent nature of `Admin::close`, it's fine to close it from both the `AbstractWorkerSourceTask` and `ConnectorOffsetBackingStore` classes.

The underlying point about ownership being unclear is fair, though, and the control flow that creates the topic admin in this method is especially convoluted. I've added a couple comments that hopefully clear things up. I wish the logic itself could be better but as it is, things are easy to test and that's taken advantage of by adding a ton of coverage in the [WorkerTest suite](https://github.com/C0urante/kafka/blob/fe675683b4671714d02c8e674bd36347d0c345e9/connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerTest.java#L1592-L1653). Happy to refactor if there's an easily-testable approach that's clearer, though.

tombentley replied:
The conditions under which we actually end up calling `topicAdminCreator.get()` are:

```java
sourceConfig.offsetsTopic() != null && config.connectorOffsetsTopicsPermitted() // connector-specific store
|| config.topicCreationEnable() && sourceConfig.usesTopicCreation() // task needs for topic creation
```

These are the only circumstances in which `topicAdmin.get()` would return non-null.

So I wonder if we should just put that logic directly in `doBuild()`, and have `offsetStoreForRegularSourceTask()` accept a nullable `topicAdminSupplier` (which will be guaranteed non-null in the case where it's actually called).

Wdyt?

tombentley replied:
Looking at the tests it doesn't seem we have any that assert that the `Admin` actually gets closed. `WorkerSourceTaskTest` asserts this, but for this code we're reliant that every invocation of `doBuild` always returns a `WorkerSourceTask` which is always stopped. I can see this is pretty hard to test though.

C0urante replied:
👍 I added the logic for checking those conditions to `doBuild`. It's still a little clunky since we're checking `regularSourceTaskUsesConnectorSpecificOffsetsStore` in two places now (`doBuild` and `offsetStoreForRegularSourceTask`), but for testing purposes it's easier to work with since we can continue to unit test `offsetStoreForRegularSourceTask` independently without having to go through `doBuild`. And it's certainly more readable than the caching `Supplier` logic that it replaced.

Good point about ensuring that we stop `WorkerSourceTask` instances, we definitely do a poor job of that when we fail to build a task. In fact, this has already been documented as a bug with [an accompanying fix PR](https://github.com/apache/kafka/pull/11608). If that's worth pursuing at some point let me know and I can rebase onto the latest trunk.

tombentley replied:
Thanks! When you have time it would be great if you could rebase that PR.

C0urante replied:
👍 done.

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java:
```diff
@@ -16,56 +16,322 @@
+            @Override
+            public boolean cancel(boolean mayInterruptIfRunning) {
+                return workerGetFuture.cancel(mayInterruptIfRunning)
+                        || connectorGetFuture.cancel(mayInterruptIfRunning);
```
Is it correct that if `workerGetFuture.cancel` was already cancelled, and returned true, then we won't try to cancel `connectorGetFuture`?

C0urante replied:
Ah yes, good point. Switched to `|` instead of `||`, which disables short-circuiting so that it's guaranteed that `cancel` is called on both futures.

C0urante replied:
Also added a comment to this part just in case someone sees that and things that a single pipe instead of a double pipe is just a typo.

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java:
```diff
@@ -16,56 +16,322 @@
+            public Map<ByteBuffer, ByteBuffer> get(long timeout, TimeUnit unit) throws InterruptedException, ExecutionException, TimeoutException {
+                long endTime = time.milliseconds() + unit.toMillis(timeout);
+                Map<ByteBuffer, ByteBuffer> result = new HashMap<>(workerGetFuture.get(timeout, unit));
+                timeout = Math.max(1, endTime - time.milliseconds());
```
It's quite confusing that `timeout` is initially in `unit`, but in this reassignment becomes ms. I think maybe a new `timeoutRemainingMs` would be clearer.

C0urante replied:
👍 done.

tombentley reviewed (commented):
Thanks @C0urante! This is a first pass. I haven't looked at tests yet.

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java:
```diff
@@ -63,8 +67,62 @@
+     * @return an offset store backed by the given topic and Kafka clients
+     */
+    public static KafkaOffsetBackingStore forTask(String topic,
+                                                       Producer<byte[], byte[]> producer,
+                                                       Consumer<byte[], byte[]> consumer,
+                                                       TopicAdmin topicAdmin) {
```
nit: indent

C0urante replied:
Ack, addressed

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java:
```diff
@@ -63,8 +67,62 @@
+     * @return a read-only offset store backed by the given topic and Kafka clients
+     */
+    public static KafkaOffsetBackingStore forConnector(String topic,
+                                                  Consumer<byte[], byte[]> consumer,
+                                                  TopicAdmin topicAdmin) {
```
nit: indent

C0urante replied:
Ack, addressed

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/KafkaBasedLog.java:
```diff
@@ -381,6 +435,10 @@ Map<TopicPartition, Long> readEndOffsets(Set<TopicPartition> assignment, boolean
             } catch (UnsupportedVersionException e) {
                 // This may happen with really old brokers that don't support the auto topic creation
                 // field in metadata requests
+                if (requireAdminForOffsets) {
+                    // Should be handled by the caller during log startup
+                    throw e;
```
Should we set `admin = null` in this case, like the comment said:
`// Forget the reference to the admin so that we won't even try to use the admin the next time this method is called` ?

C0urante replied:
We want to keep throwing the exception here if the method is called again; nullifying the `admin` field would stop that from happening. That does remind me though, we should probably copy the same error message from [here](https://github.com/apache/kafka/blob/f421c008aa0495441ce6f017cdd041937a26d10f/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaConfigBackingStore.java#L333-L338) into the code path for starting connector offset stores. I've done that in my latest push; LMKWYT.

showuon replied:
SGTM. Thanks.

showuon commented on connect/runtime/src/test/java/org/apache/kafka/connect/runtime/WorkerTest.java:
```diff
@@ -1279,6 +1288,455 @@ public void testAdminConfigsClientOverridesWithNonePolicy() {
+        workerProps.remove("producer." + TRANSACTIONAL_ID_CONFIG);
+        connectorProducerOverrides.put(BOOTSTRAP_SERVERS_CONFIG, "localhost:489");
+        connectorProducerOverrides.put(ENABLE_IDEMPOTENCE_CONFIG, "false");
+        connectorProducerOverrides.put(TRANSACTIONAL_ID_CONFIG, "yet-another-transactional-id");
```
Could user overrides the `transactional.id` to `null`? Should we add this test case?

C0urante replied:
I don't _think_ that can happen without causing other issues, since the user-supplied configs have come from either the JSON REST API or the Java properties worker file. The former doesn't deserializes things into a literal null and although the latter does, as of https://github.com/apache/kafka/pull/11333, we fail validation for connector configs in that case.

But, given how little additional effort it is, I don't see a problem with adding that check. At the very least, it helps prevent a 500 instead of a 400 response during config validation.

showuon reviewed (commented):
I've made a pass, left some comments. Thanks.

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/TopicAdmin.java:
```diff
@@ -274,18 +276,23 @@ public static NewTopicBuilder defineTopic(String topicName) {
+     * @param bootstrapServers the Kafka cluster targeted by the admin
+     * @param adminClient the {@link Admin} to use under the hood
+     */
+    public TopicAdmin(Object bootstrapServers, Admin adminClient) {
```
This constructor kinda confuses the ownership of the `Admin` client. I think things are cleaner when the TopicAdmin instantiates (and thus owns the `admin`). Note, it looks like there are no callers for `TopicAdmin.admin`.  It seems that the call sites in `doBuild` could simply pass the map of configs (and the `bootstrapServers` looked up from that), rather than instantiating the `admin` and then passing it to the TopicAdmin.

Obviously the test code has slightly different requirements, meaning we still need this constructor. I did also wonder whether we could also get rid of `bootstrapServers` by defining `toString` on `KafkaAdminClient` and using that for the logging and exceptions here in `TopicAdmin`. Perhaps that's worth a followup PR at some point, (though perhaps there are benefits to hiding bootstrap servers from receivers of clients).

C0urante replied:
Good points. I've:

- Demoted the visibility of the constructor that accepts an `Admin` instance back to package-private
- Documented that it's the caller's responsibility to close the topic admin in order to release the resources allocated for its `Admin`
- Removed the unused `admin` method
- Switched over to the `TopicAdmin(Map<String, Object>)` constructor in the `Worker` class

I originally did this refactoring because I was managing the lifecycle of the `Admin` differently, but kept part of it because it made testing much easier when all you had to pass in was an `Admin` and a `String` instead of an `Admin` and an entire config map. It also seemed like an anti-pattern to have a constructor that accepted both an already-instantiated client object, and the configuration for that client object.

C0urante commented:
Hey guys--thanks for the reviews, really appreciate the rapid responses here. I found a bug that's been a bit trickier to solve than expected and have had little time to work on it this week. I plan to push the next draft by Friday at the very latest.

If it matters, the bug is that the offset stores for regular (non-exactly-once) source tasks, and source connectors, are never started. I'm planning on fixing that first, then adding an integration test case to https://github.com/apache/kafka/pull/11782 to simulate a soft downgrade where someone disables exactly-once support on their worker after creating a connector and letting it run for a bit, and finally, manually auditing the changes for KIP-618 to catch any other potential bugs related to improper initialization or cleanup of resources.

C0urante commented:
Pushed the next draft which should fix the offset store startup bug and address outstanding review comments.

C0urante commented:
I've completed the remaining follow-up tasks.
### Audit of resource allocation and cleanup:
- Documented ownership expectations more clearly with the `KafkaBasedLog` and `KafkaOffsetBackingStore` classes, especially with regards to their underlying Kafka clients
- Found and fixed a small bug in `ExactlyOnceWorkerSourceTask` where its `transactionMetrics` field was never closed
- Found a long-lived bug where `HeaderConverter` instances are never closed; given that this was not introduced by any of the changes for KIP-618, I've filed https://issues.apache.org/jira/browse/KAFKA-14007 and would like to address it separately
- Found no other resource allocation, initialization, or cleanup bugs in the code paths related to KIP-618

### Integration testing
- Added a new test to https://github.com/apache/kafka/pull/11782 (or rather, augmented an existing one) for running a source connector with a separate offsets topic with exactly-once support disabled

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java:
```diff
@@ -16,56 +16,325 @@
-        // TODO
+        // Worker offset store should not be stopped as it may be used for multiple connectors
+        connectorStore.ifPresent(OffsetBackingStore::stop);
+        connectorStoreAdmin.ifPresent(TopicAdmin::close);
```
Given that we close the `TopicAdmin` for source tasks in `AbstractWorkerSourceTask::close` [here](https://github.com/apache/kafka/blob/1e21201ea24389bdaccb8a462f3a53e356b58a58/connect/runtime/src/main/java/org/apache/kafka/connect/runtime/AbstractWorkerSourceTask.java#L306-L308), this may seem redundant. However, I've chosen to keep it in for two reasons:
1. The `WorkerConnector` class has no access to the `TopicAdmin` used for its offset store; we'd have to add another constructor parameter to that class and manually close the `TopicAdmin` during shutdown if we removed this line
2. It seems safer to leave this in. Although `Autocloseable::close` is not required to be idempotent, it is strongly recommended, and since all current calls to `ConnectorOffsetBackingStore::stop` are wrapped in `Utils::closeQuietly` anyways, the risks are low. On the other hand, by leaving this out, we would make it easier to introduce a resource leak in the future, which would have more serious consequences.

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -1425,6 +1426,256 @@ public WorkerTask doBuild(Task task,
+        // a network ping to a remote Kafka cluster inside the herder's tick thread (which is where this
+        // logic takes place right now) in case that takes a while.
+        return offsetsTopic.equals(config.offsetsTopic())
+                && config.bootstrapServers().equals(producerProps.get(BOOTSTRAP_SERVERS_CONFIG));
```
I guess we could improve this slightly by not comparing as strings, but as `Set<String>`. That way at least the order wouldn't matter, which it looks like it does right now.

I've wondered whether we should add some kind of `cluster.id` to the client configs which could be used to assert that the client was connected to the expected cluster. If we did do that, and they were specified in the configs here, then you could safely know at this point whether they were the same cluster without needing to connect at this point.

C0urante replied:
True, we can make this order-independent by comparing sets. Not too much work, so I've done that.

I have a long-term vision of relocating this logic into the `WorkerTask` class (or its subclasses) so that we can ping the remote Kafka cluster, grab its ID there, and decide at that point whether or not we need to construct an additional offsets store, all without having to worry about blocking up the herder tick thread. This would obviate the need for users to do any additional work at all, but it would require a significant refactoring of this part of the code base, so I've held off on it for now.

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java:
```diff
@@ -131,7 +203,24 @@ private KafkaBasedLog<byte[], byte[]> createKafkaBasedLog(String topic, Map<Stri
+        return new KafkaBasedLog<>(topic, producerProps, consumerProps, adminSupplier, consumedCallback, Time.SYSTEM, createTopics);
+    }
+
+    protected NewTopic topicDescription(final String topic, final WorkerConfig config) {
```
Might `newTopicSpecification` be a better name? It's just the "description" is a bit ambiguous: it could be describing an _existing_ topic.

C0urante replied:
Ack, done

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java:
```diff
@@ -142,16 +231,40 @@ private KafkaBasedLog<byte[], byte[]> createKafkaBasedLog(String topic, Map<Stri
+            if (exactlyOnce) {
+                message = "Enabling exactly-once support for source connectors requires a Kafka broker version that allows "
+                        + "admin clients to read consumer offsets. Please either disable the worker's exactly-once "
+                        + "support for source connectors, or use a newer Kafka broker version.";
```
```suggestion
                        + "support for source connectors, or upgrade to a newer Kafka broker version.";
```

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java:
```diff
@@ -142,16 +231,40 @@ private KafkaBasedLog<byte[], byte[]> createKafkaBasedLog(String topic, Map<Stri
+                message = "When " + ConsumerConfig.ISOLATION_LEVEL_CONFIG + "is set to "
+                        + IsolationLevel.READ_COMMITTED.name().toLowerCase(Locale.ROOT)
+                        + ", a Kafka broker version that allows admin clients to read consumer offsets is required. "
+                        + "Please either reconfigure the worker or connector, or use a newer Kafka broker version.";
```
```suggestion
                        + "Please either reconfigure the worker or connector, or upgrade to a newer Kafka broker version.";
```

tombentley commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java:
```diff
@@ -142,16 +231,40 @@ private KafkaBasedLog<byte[], byte[]> createKafkaBasedLog(String topic, Map<Stri
+     * Stop reading from and writing to the offsets topic, and relinquish resources allocated for interacting
+     * with it, including Kafka clients.
+     * <p>
+     * <b>Note:</b> if the now-deprecated {@link #KafkaOffsetBackingStore()} constructor was used to create
+     * this store, the underlying admin client allocated for interacting with the offsets topic will be closed.
+     * On the other hand, if the recommended {@link #KafkaOffsetBackingStore(Supplier)} constructor was used to
+     * create this store, the admin client derived from the given {@link Supplier} will not be closed and it is the
+     * caller's responsibility to manage its lifecycle accordingly.
```
Great doc, but maybe we should also have some on those constructors themselves?

C0urante replied:
👍 good point, done

tombentley reviewed (approved):
Thanks @C0urante, I left a few more comment, but they're nits at this stage.

C0urante commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerConnector.java:
```diff
@@ -281,7 +283,9 @@ void doShutdown() {
             Utils.closeQuietly(metrics, "connector metrics for " + connName);
             Utils.closeQuietly(offsetStorageReader, "offset reader for " + connName);
-            Utils.closeQuietly(offsetStore::stop, "offset backing store for " + connName);
+            if (offsetStore != null) {
```
Discovered this bug while working on https://github.com/apache/kafka/pull/12307 for https://issues.apache.org/jira/browse/KAFKA-14006.

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/Worker.java:
```diff
@@ -1425,6 +1428,268 @@ public WorkerTask doBuild(Task task,
+            // Should never happen by this point, but if it does, make sure to present a readable error message to the user
+            throw new ConnectException("Failed to parse bootstrap servers property in producer config", e);
+        }
+        producerBootstrapServers = new HashSet<>(producerBootstrapServers);
```
We've add all bootstrap servers into `producerBootstrapServers` set in L1672, why should we re-create another hashSet here?

C0urante replied:
Good catch, left in accidentally. Removed

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java:
```diff
@@ -16,56 +16,325 @@
     }
 
+    /**
+     * If configured to use a connector-specific offset store, {@link OffsetBackingStore#start() stop} that store,
```
OffsetBackingStore#start() -> OffsetBackingStore#stop() ?

C0urante replied:
Good catch, done

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java:
```diff
@@ -16,56 +16,325 @@
 
+    /**
+     * If configured to use a connector-specific offset store, {@link OffsetBackingStore#start() stop} that store,
+     * and the {@link TopicAdmin} used by that store.
```
and [close] the {@link TopicAdmin} used by that store.

C0urante replied:
Ack, done

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/ConnectorOffsetBackingStore.java:
```diff
@@ -16,56 +16,325 @@
+        return primaryStore.set(values, (primaryWriteError, ignored) -> {
+            if (secondaryStore != null) {
+                if (primaryWriteError != null) {
+                    log.trace("Skipping offsets write to secondary store because primary write has failed", primaryWriteError);
```
Why did we log `trace` to this primary store error, but `warn` to the secondary store error in L290?

C0urante replied:
The failure to write to the primary store will probably be logged elsewhere at a higher level, since we report it to the callback. We swallow failures to write to the secondary store, so this is the only chance we get to log those.

showuon replied:
I see. Thanks.

showuon commented on connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java:
```diff
@@ -63,28 +70,109 @@
         if (topic == null || topic.trim().length() == 0)
             throw new ConfigException("Offset storage topic must be specified");
 
+        exactlyOnce = config.exactlyOnceSourceEnabled();
```
unused variable?

C0urante replied:
Used [here](https://github.com/C0urante/kafka/blob/ef5a43b70bc1172636cce43de0fcf493825da24d/connect/runtime/src/main/java/org/apache/kafka/connect/storage/KafkaOffsetBackingStore.java#L258-L267).

Added `this.` on writes to this field and others in the `configure` method to make it easier to identify instance vs. function variables.

showuon reviewed (commented):
Made another pass, left some minor comments. Overall LGTM. During the review, I found a bug [KAFKA-14012](https://issues.apache.org/jira/browse/KAFKA-14012). But it's not related to this change and can be fixed in another PR. Thanks.

showuon reviewed (approved):
LGTM!
